zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java

```diff
@@ -372,6 +400,7 @@ protected void pRequest2Txn(int type, long zxid, Request request, Record record,
             if (nodeRecord.childCount > 0) {
                 throw new KeeperException.NotEmptyException(path);
             }
+            ephemeralNodesPerSessionMap.computeIfPresent(nodeRecord.stat.getEphemeralOwner(), (key, value) -> value - 1);
```

Comment:
I think this class `PrepRequestProcessor` may not be the correct place to do the ephemeral node number change update, since in this class we put all the outstanding requests into a queue to wait for processing, but not actually process the requests in this class. You can explore `FinalRequestProcessor` to see if that is the more appropriate place. We can discuss about it offline.